### PR TITLE
Docker update

### PR DESCRIPTION
**Pull request by MiXaiLL76**
Hello friend! I see this repository is frozen in time. Perhaps I will revive it a little.

Your ENV for running the project is very outdated, but some people still want to use it. I solved this problem in a simple and concise way - with the help of Docker.

In simple words, now you can build your code on any hardware and it will work as intended

### CODE REVIEW

File: src/main.cpp [hunk 1]
```diff
@@ -1,13 +1,7 @@
 #include "stdafx.h"
-#define WITH_GUI 1
```
Comment (bmpix) [inline src/main.cpp:2]: i would like to keep that line for those who want GUI.

File: src/main.cpp [hunk 2]
```diff
@@ -1,13 +1,7 @@
...
-#ifdef _WIN32
```
Comment (bmpix) [inline src/main.cpp:3]: perhaps #ifdef QT_IS_AVAILABLE ?

File: src/main.cpp [hunk 3]
```diff
@@ -263,9 +257,7 @@ std::string type2str(int type) {
 
 int main(int argc, char *argv[])
 {
-#ifdef _WIN32
```
Comment (bmpix) [inline src/main.cpp:266]: same, #ifdef QT_IS_AVAILABLE?

File: src/main.cpp [hunk 4]
```diff
@@ -548,9 +534,9 @@ int main(int argc, char *argv[])
 #ifdef WITH_GUI
 	mw.setVectorization("Final", newVectorization);
 #endif
-#ifdef _WIN32
```
Comment (bmpix) [inline src/main.cpp:551]: same as before

**Review by bmpix — CHANGES_REQUESTED** [review]
Thanks for the pull request! I'm happy to merge once we support all three options: no qt, qt with GUI, and qt without gui. I think it's just fixing the ifdefs

File: src/main.cpp (hunk 1 above)
Comment (MiXaiLL76) [reply src/main.cpp:2]: Why? This variable can be set from the compiler.

target_compile_definitions(polyvector_thing PRIVATE WITH_GUI=1)

As it already works. By default with GUI

**Comment by MiXaiLL76** [comment]
Divided the library into 3 parts. CLI \ CLI QT \ GUI QT.

You can see the compilation examples in the dockerfile. But in short - I added the cmake variables WITH_GUI & WITH_QT.

I can provide objdump as confirmation.

CLI
```bash
Version References:
  required from libgcc_s.so.1:
    0x0b792650 0x00 14 GCC_3.0
    0x09275a60 0x00 13 GCC_4.0.0
  required from libgomp.so.1:
    0x042623d0 0x00 16 GOMP_4.0
    0x04262440 0x00 10 OMP_1.0
  required from libm.so.6:
    0x09691a75 0x00 06 GLIBC_2.2.5
  required from libstdc++.so.6:
    0x0bafd178 0x00 15 CXXABI_1.3.8
    0x02297f89 0x00 12 GLIBCXX_3.4.9
    0x0297f861 0x00 11 GLIBCXX_3.4.11
    0x0297f865 0x00 08 GLIBCXX_3.4.15
    0x0297f871 0x00 07 GLIBCXX_3.4.21
    0x056bafd3 0x00 04 CXXABI_1.3
    0x08922974 0x00 03 GLIBCXX_3.4
  required from libc.so.6:
    0x09691974 0x00 17 GLIBC_2.3.4
    0x0d696914 0x00 09 GLIBC_2.4
    0x06969194 0x00 05 GLIBC_2.14
    0x09691a75 0x00 02 GLIBC_2.2.5
```

CLI QT
```bash
Version References:
  required from libgomp.so.1:
    0x042623d0 0x00 19 GOMP_4.0
    0x04262440 0x00 17 OMP_1.0
  required from libQt5Core.so.5:
    0x058a2817 0x00 16 Qt_5.7
    0x00058a25 0x00 12 Qt_5
  required from libgcc_s.so.1:
    0x0b792650 0x00 13 GCC_3.0
    0x09275a60 0x00 10 GCC_4.0.0
  required from libm.so.6:
    0x09691a75 0x00 08 GLIBC_2.2.5
  required from libc.so.6:
    0x06969194 0x00 15 GLIBC_2.14
    0x09691974 0x00 14 GLIBC_2.3.4
    0x0d696914 0x00 07 GLIBC_2.4
    0x09691a75 0x00 04 GLIBC_2.2.5
  required from libstdc++.so.6:
    0x0bafd178 0x00 18 CXXABI_1.3.8
    0x02297f89 0x00 11 GLIBCXX_3.4.9
    0x0297f861 0x00 09 GLIBCXX_3.4.11
    0x0297f871 0x00 06 GLIBCXX_3.4.21
    0x0297f865 0x00 05 GLIBCXX_3.4.15
    0x056bafd3 0x00 03 CXXABI_1.3
```

GUI QT
```bash
Version References:
  required from libgomp.so.1:
    0x042623d0 0x00 21 GOMP_4.0
    0x04262440 0x00 19 OMP_1.0
  required from libgcc_s.so.1:
    0x0b792650 0x00 15 GCC_3.0
    0x09275a60 0x00 13 GCC_4.0.0
  required from libm.so.6:
    0x09691a75 0x00 11 GLIBC_2.2.5
  required from libQt5Core.so.5:
    0x058a2817 0x00 18 Qt_5.7
    0x00058a25 0x00 07 Qt_5
  required from libc.so.6:
    0x06969194 0x00 17 GLIBC_2.14
    0x09691974 0x00 16 GLIBC_2.3.4
    0x0d696914 0x00 10 GLIBC_2.4
    0x09691a75 0x00 06 GLIBC_2.2.5
  required from libQt5Gui.so.5:
    0x00058a25 0x00 04 Qt_5
  required from libstdc++.so.6:
    0x0bafd178 0x00 20 CXXABI_1.3.8
    0x02297f89 0x00 14 GLIBCXX_3.4.9
    0x0297f861 0x00 12 GLIBCXX_3.4.11
    0x0297f871 0x00 09 GLIBCXX_3.4.21
    0x0297f865 0x00 08 GLIBCXX_3.4.15
    0x056bafd3 0x00 05 CXXABI_1.3
    0x08922974 0x00 03 GLIBCXX_3.4
  required from libQt5Widgets.so.5:
    0x00058a25 0x00 02 Qt_5
```

**Comment by MiXaiLL76** [comment]
I accidentally saw your project thanks to friends and realized that people use it.
The current implementation with a docker image is necessary if you are not going to maintain the code in the future and do not want to switch to more modern versions of libraries.

In theory, we can discuss it separately, perhaps I could help with the implementation of your code "as a library" for python3. You can take a look at my personal project, I am implementing something similar there https://github.com/MiXaiLL76/faster_coco_eval/tree/mask_cpp

**Comment by bmpix** [comment]
Merged. Thanks! And yes, library for python3 would be awesome.

**Pull request merged**